### PR TITLE
TEST-#5943: Rebalance tests between different CI jobs

### LINKED ISSUE #5943: `test-all` job takes too much time in CI
`test-all` job takes from [1 hour (for Ray)](https://github.com/modin-project/modin/actions/runs/4613660639/jobs/8155949377?pr=5939) up to[ 2 hours (for Dask)](https://github.com/modin-project/modin/actions/runs/4613660639/jobs/8155949459?pr=5939).

We should consider splitting this job into multiple ones so they can be run in parallel

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

This PR splits some CI jobs into multiple ones so they can be run in parallel and take less time overall.

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #5943 
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -850,35 +884,13 @@ jobs:
...
-          - modin/numpy/test/test_array_indexing.py
-          - modin/numpy/test/test_array_math.py
+        test_task:
+          - group_1
+          - group_2
+          - group_3
     env:
       MODIN_ENGINE: ${{matrix.engine}}
-    name: test-windows
+    test-ubuntu (engine ${{matrix.engine}}, python ${{matrix.python-version}}, ${{matrix.test_task}})
```
Comment (anmyachev) [inline .github/workflows/ci.yml:893]:
```suggestion
    name: test-windows (engine ${{matrix.engine}}, python ${{matrix.python-version}}, ${{matrix.test_task}})
```

**Comment by dchigarev** [comment]
So I tried to experiment in a separate branch (#5937) in order to get rid of if-elsing at every job's step and unfortunately found no success:

Initially, I thought that I can define one more matrix variable that would contain different testing stages and then loop across the testing files to execute the `- run:` command on them:
```yaml

matrix:
    commands: [ # define multiple chunks of tests
        [pytest test_dataframe.py, pytest test_groupby.py, ...],
        [pytest test_binary.py, pytest test_series.py, ...],
        ...
    ]
...
steps:
    for command in ${{matrix.commands[@]}}; do
        - run: $command
    end
```
I soon found out that it is impossible to dynamically generate steps that way.


Then I decided to try executing several tests inside of a single `- run:` statement hoping that it would somehow separate the output for each `eval` command:
```yaml
matrix:
    commands: [ # define multiple chunks of tests
        ["pytest test_dataframe.py; pytest test_groupby.py; ..."],
        ["pytest test_binary.py; pytest test_series.py; ..."],
        ...
    ]
...
steps:
    - run: |
        readarray -d ';' commands_arr <<< "${{ matrix.command }}" # split the command string by separator ';'
        for test in "${commands_arr[@]}"; do
            eval "$test" # execute test
        done
```
Unfortunately, it doesn't anyhow separate either inputs or outputs. So the command to execute several tests is shown as a single call and the output is simply the concatenation of the outputs of each test without any separation:

image

I tried to find a way to somehow indicate to gh-actions the commands that I want to be visualized separately in the UI and only found [this post from stackoverflow](https://stackoverflow.com/questions/71020739/dynamic-github-actions-steps).

@anmyachev Given this, I think it's reasonable to proceed with your current PR as it is, unless we will meet a reviewer who's more experienced in gh-actions and CI stuff.

If you don't mind then I can take ownership of your PR, and make it ready for review.

**Comment by anmyachev** [comment]
> So I tried to experiment in a separate branch (#5937) in order to get rid of if-elsing at every job's step and unfortunately found no success:
> 
> Initially, I thought that I can define one more matrix variable that would contain different testing stages and then loop across the testing files to execute the `- run:` command on them:
> 
> ```yaml
> matrix:
>     commands: [ # define multiple chunks of tests
>         [pytest test_dataframe.py, pytest test_groupby.py, ...],
>         [pytest test_binary.py, pytest test_series.py, ...],
>         ...
>     ]
> ...
> steps:
>     for command in ${{matrix.commands[@]}}; do
>         - run: $command
>     end
> ```
> 
> I soon found out that it is impossible to dynamically generate steps that way.
> 
> Then I decided to try executing several tests inside of a single `- run:` statement hoping that it would somehow separate the output for each `eval` command:
> 
> ```yaml
> matrix:
>     commands: [ # define multiple chunks of tests
>         ["pytest test_dataframe.py; pytest test_groupby.py; ..."],
>         ["pytest test_binary.py; pytest test_series.py; ..."],
>         ...
>     ]
> ...
> steps:
>     - run: |
>         readarray -d ';' commands_arr <<< "${{ matrix.command }}" # split the command string by separator ';'
>         for test in "${commands_arr[@]}"; do
>             eval "$test" # execute test
>         done
> ```
> 
> Unfortunately, it doesn't anyhow separate either inputs or outputs. So the command to execute several tests is shown as a single call and the output is simply the concatenation of the outputs of each test without any separation:
> 
> image
> 
> I tried to find a way to somehow indicate to gh-actions the commands that I want to be visualized separately in the UI and only found [this post from stackoverflow](https://stackoverflow.com/questions/71020739/dynamic-github-actions-steps).
> 
> @anmyachev Given this, I think it's reasonable to proceed with your current PR as it is, unless we will meet a reviewer who's more experienced in gh-actions and CI stuff.
> 
> If you don't mind then I can take ownership of your PR and make it ready for review.

@dchigarev ok, no problem :) we at least tried

**Review by mvashishtha — APPROVED** [review]
thanks so much @dchigarev @anmyachev ! 

I originally thought we'd have an enormous matrix with a different job for each test. Why not go with that approach? I suppose it would be hard to read all the results, and we get a higher chance of flaking on at least one job.

**Comment by dchigarev** [comment]
> I originally thought we'd have an enormous matrix with a different job for each test. Why not go with that approach?

There's a significant overhead to run each job. For example, notice that in this case we spend 40% of the time just to setup a conda environment and only 60% to actually run the tests:
https://github.com/modin-project/modin/actions/runs/4620689768/jobs/8171403927?pr=5890

So, splitting the jobs further would result into that we spend 70%+ of the CI time just setting up conda environments, not to mention that it would dramatically increase the number of jobs themselves, which may result into that significant part of them would be delayed because of long queues.

**Review by vnlitvinov — APPROVED** [review]

**Review by YarShev — APPROVED** [review]

**Pull request merged**